org.jacoco.core/src/org/jacoco/core/internal/instr/Companions.java

```diff
@@ -0,0 +1,197 @@
+
+	static class Companion {
+
+		final WeakReference<Class<?>> cls;
+
+		int usedFields;
+
+		Companion(Class<?> cls) {
+			// TODO(Godin):
+			// we can utilize ReferenceQueue to cleanup unused probe arrays
+			// https://github.com/jacoco/jacoco/issues/134
+			this.cls = new WeakReference<Class<?>>(cls);
+		}
+
+	}
+
+	private final Map<ClassLoader, Companion> companions = new WeakHashMap<ClassLoader, Companion>();
+
+	private final RuntimeData runtimeData;
+
+	public Companions(final RuntimeData runtimeData) {
+		this.runtimeData = runtimeData;
+	}
+
+	/**
+	 * Creates instrumented version of the given class.
+	 *
+	 * @param classLoader
+	 *            defining loader
+	 * @param className
+	 *            name of the class
+	 * @param buffer
+	 *            definition of the class
+	 * @return instrumented definition
+	 */
+	public byte[] instrument(final ClassLoader classLoader,
+			final String className, final byte[] buffer) {
+		final boolean java9 = Java9Support.isPatchRequired(buffer);
+		final ClassReader reader = new ClassReader(
+				java9 ? Java9Support.downgrade(buffer) : buffer);
+
+		final int probeCount = ProbeArrayStrategyFactory.getProbeCounter(reader)
+				.getCount();
+		if (probeCount == 0) {
+			return buffer;
+		}
+		final long classId = CRC64.checksum(buffer);
+		final IProbeArrayStrategy strategy = createStrategyFor(classLoader,
+				classId, className, probeCount);
+		final ClassWriter writer = new ClassWriter(reader, 0);
+		final ClassVisitor visitor = new ClassProbesAdapter(
+				new ClassInstrumenter(strategy, writer), true);
+		reader.accept(visitor, ClassReader.EXPAND_FRAMES);
+
+		final byte[] result = writer.toByteArray();
+		if (java9) {
+			Java9Support.upgrade(result);
+		}
+		return result;
+	}
+
+	private IProbeArrayStrategy createStrategyFor(final ClassLoader classLoader,
+			final long classId, final String className, final int probeCount) {
+		final Class<?> cls;
+		final String field;
+		synchronized (companions) {
+			Companion companion = companions.get(classLoader);
+			if (companion == null) {
+				final byte[] definition = createCompanionClass();
+				try {
+					cls = (Class<?>) DEFINE_CLASS.invoke(classLoader,
+							COMPANION_NAME, definition, 0, definition.length);
+					companion = new Companion(cls);
+				} catch (IllegalAccessException e) {
+					throw new RuntimeException(e);
+				} catch (InvocationTargetException e) {
+					throw new RuntimeException(e);
+				}
+				companions.put(classLoader, companion);
+			} else {
+				cls = companion.cls.get();
+				// if class loader is alive, then class is also alive
+				assert cls != null;
+			}
+
+			if (companion.usedFields == FIELDS_PER_CLASS) {
+				throw new UnsupportedOperationException();
+			}
+			field = "p" + companion.usedFields;
+			companion.usedFields++;
+		}
+
+		final boolean[] probes = runtimeData
+				.getExecutionData(classId, className, probeCount).getProbes();
+		try {
+			cls.getField(field).set(null, probes);
```

Comment:
@bjkail you right that `Class.getField` is O(n) , what also explains degradation of performance, which was observed when `usedFields` was stored inside of "companion class". Implemented both `tableswitch` and generation of multiple classes.